crates/marketplace/src/testing/order_test.rs

```diff
@@ -0,0 +1,527 @@
+};
+use hotshot::rand;
+use std::time::Duration;
+
+/// The function checks whether the common part of two transaction vectors have the same order
+fn order_check<T: Eq + Clone + Debug>(
+    transaction_history: Vec<T>,
+    all_transactions: Vec<Vec<T>>,
+) -> bool {
+    let all_transactions_vec = all_transactions.into_iter().flatten().collect::<Vec<_>>();
+    tracing::debug!(
+        "Doing order check, transaction_history = {:?}, all_transactions = {:?}",
+        transaction_history,
+        all_transactions_vec
+    );
+    let common_txs: Vec<_> = transaction_history
+        .iter()
+        .filter(|item| all_transactions_vec.contains(item))
+        .collect();
+
+    let another_common_txs: Vec<_> = all_transactions_vec
+        .iter()
+        .filter(|item| transaction_history.contains(item))
+        .collect();
+
+    common_txs == another_common_txs
+}
+
+/// This test simulates multiple builder states receiving messages from the channels and processing them
+/// and focus specifically on orders.
+/// It's fine that leader doesn't include some of transactions we've given, or interspersed with other transactions,
+/// as long as the order is correct it will be good.
+#[async_std::test]
+async fn test_builder_order() {
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+    tracing::info!("Testing the builder core with multiple messages from the channels");
+
+    // Number of views to simulate
+    const NUM_ROUNDS: usize = 10;
+    // Number of transactions to submit per round
+    const NUM_TXNS_PER_ROUND: usize = 5;
+    // Capacity of broadcast channels
+    const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+    // Number of nodes on DA committee
+    const NUM_STORAGE_NODES: usize = 4;
+
+    let (senders, global_state) = start_builder_state(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+    // Transactions to send
+    let all_transactions = (0..NUM_ROUNDS)
+        .map(|round| {
+            (0..NUM_TXNS_PER_ROUND)
+                .map(|tx_num| TestTransaction::new(vec![round as u8, tx_num as u8]))
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    // generate three different random number between (0..NUM_ROUNDS) to do some changes for output transactions
+    let mut unique_rounds = HashSet::new();
+    #[deny(clippy::modulo_one)]
+    while unique_rounds.len() < 3 {
+        let random_round = std::cmp::max(rand::random::<usize>() % (NUM_ROUNDS - 2), 1);
+        unique_rounds.insert(random_round);
+    }
+    let random_rounds: Vec<_> = unique_rounds.into_iter().collect();
+    let skip_round = random_rounds[0]; // the round we want to skip all the transactions
+    let adjust_add_round = random_rounds[1]; // the round we want to randomly add some transactions
+    let adjust_remove_tail_round = random_rounds[2]; // the round we want to cut off the end of the bundle
```

Comment:
Because we rolled these rounds separately without any consideration on each other, we may end up with some of these triggers on the same round.  In the case of the `skip_round` we might end up skipping the other round consequencers if they rolled the same round.

Reply:
If I understand @Ayiga's comment correctly, the potential issue here is that `skip_round` may be the same as the `adjuct_*` round, which will cause the latter to be skipped, i.e., we may end up not testing the `adjust_*` case.

Reply:
`skip_round` is guaranteed to be different from `adjust_round` because I'm using a unique_rounds hash set to keep track of all the unique random numbers (refer to line 77 - 83). Or am I missing something?

Reply:
I see! I think this is correct, but a bit complicated. Could we simplify it like this:
```
let range = (1..8).collect();

let chosen_numbers = range.choose_multiple(&mut rng, 3).cloned().collect();
```

Here `choose_multiple` is from `IteratorRandom`. There may be other ways to implement this. I just feel like we don't have to use a `HashSet` and do more than 3 times of randomization for this.